File: src/listener_zmq.cpp
/**
* listener_zmq.cpp
* Author: Ravi Joshi
* Date: 2018/01/18
*/

#include <ros/ros.h>
#include <zmq.hpp>

int main(int argc, char** argv)
{
    ros::init(argc, argv, "listener_zmq", ros::init_options::AnonymousName);

    zmq::context_t context(1);
    zmq::socket_t subscriber(context, ZMQ_SUB);

    std::string TOPIC = "";
    subscriber.setsockopt(ZMQ_SUBSCRIBE, TOPIC.c_str(), TOPIC.length()); // allow all messages

    int linger = 0; // Proper shutdown ZeroMQ
    subscriber.setsockopt(ZMQ_LINGER, &linger, sizeof(linger));

    subscriber.connect("tcp://localhost:6666");

    while (ros::ok())
    {
        zmq::message_t message;
        int rc = subscriber.recv(&message);
        if (rc)
        {
            std::string recv_string;

            std::istringstream iss(static_cast<char*>(message.data()));
            iss >> recv_string;
            ROS_INFO_STREAM(recv_string);
        }
    }

    // Clean up your socket and context here
    subscriber.close();
    context.close();
    return 0;
}

File: src/listener_zmq.py
#!/usr/bin/env python

import zmq
import rospy
from std_msgs.msg import String
    
def listener():
    rospy.init_node('listener_zmq', anonymous=True)

    context = zmq.Context()
    socket = context.socket(zmq.SUB)

    TOPIC=b""
    socket.setsockopt(zmq.SUBSCRIBE,TOPIC)

    linger=0
    socket.setsockopt(zmq.LINGER,linger)
    socket.connect ("tcp://localhost:6666")

    total_value = 0
    while not rospy.is_shutdown():
        string = socket.recv()
        # topic, messagedata = string.split()

        rospy.loginfo(rospy.get_caller_id() + "I heard %s", string)
    socket.close()
    context.close()

    
    
if __name__ == '__main__':
    listener()